File: ObjC-SampleApp/SideEngineSample/SideEngineSample/CustomCountDownViewController.m
//
//  CustomCountDownViewController.m
//  SideEngineSample
//
//  Created by PhoenixiOS-1 on 17/11/22.
//

#import "CustomCountDownViewController.h"

@interface CustomCountDownViewController ()

@end

@implementation CustomCountDownViewController
@synthesize mainTitleLabel,secondsLabel,countDownLabel,counter,counterTimer;

- (void)viewDidLoad {
    [super viewDidLoad];
    self.counter = [[BBSideEngineManager shared] countDownDuration];
    [self configureTimer];
}

- (IBAction)closeTapped {
    [self stopTimerWithFinished:NO];
    [[BBSideEngineManager shared] resumeSideEngine];
    if ([self isModal]) {
        [self dismissViewControllerAnimated:YES completion:nil];
    } else {
        [self.navigationController popViewControllerAnimated:YES];
    }
}

- (void)cancelAutoIncident {
    if (self.counterTimer) {
        [self.counterTimer invalidate];
        self.counterTimer = nil;
    }
    if ([self isModal]) {
        [self dismissViewControllerAnimated:YES completion:nil];
    } else {
        [self.navigationController popViewControllerAnimated:YES];
    }
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    [self.navigationController setNavigationBarHidden:YES animated:NO];
}

- (void)configureTimer {
    self.countDownLabel.text = [NSString stringWithFormat:@"%ld", (long)self.counter];
    self.counterTimer = [NSTimer scheduledTimerWithTimeInterval:1.0 target:self selector:@selector(runTimer) userInfo:nil repeats:YES];
}

- (void)openSupportScreen {
    UIApplicationState state = [UIApplication sharedApplication].applicationState;
    if (state != UIApplicationStateActive || [self isModal]) {
        [self closeTapped];
    } else {
        if (self.delegate) {
            [self.delegate didFinishTimer];
        }
        dispatch_async(dispatch_get_main_queue(), ^{
            UIStoryboard *storyboard = [UIStoryboard storyboardWithName:@"Main" bundle:nil];
            TestSideEngineSupportViewController *testViewController = [storyboard instantiateViewControllerWithIdentifier:@"TestSideEngineSupportViewController"];
            [self.navigationController pushViewController:testViewController animated:YES];
        });
    }
}

- (void)runTimer {
    self.counter--;
    self.countDownLabel.text = [NSString stringWithFormat:@"%ld", (long)self.counter];
    if (self.counter >= 0) {
        AudioServicesPlayAlertSound(kSystemSoundID_Vibrate);
    }
    if (self.counter <= 0) {
        [self stopTimerWithFinished:YES];
    }
}

- (void)stopTimerWithFinished:(BOOL)finished {
    self.countDownLabel.text = @"0";
    self.counter = 0;
    if (self.counterTimer) {
        [self.counterTimer invalidate];
        self.counterTimer = nil;
    }
    if (finished) {
        [[BBSideEngineManager shared] notifyPartner];
        dispatch_async(dispatch_get_main_queue(), ^{
            [self openSupportScreen];
        });
    }
}

- (BOOL)isModal {
    BOOL presentingIsModal = self.presentingViewController != nil;
    BOOL presentingIsNavigation = self.navigationController.presentingViewController.presentedViewController == self.navigationController;
    BOOL presentingIsTabBar = [self.tabBarController.presentingViewController isKindOfClass:[UITabBarController class]];
    return presentingIsModal || presentingIsNavigation || presentingIsTabBar;
}

@end

File: Swift-SampleApp/FlareSDKSample/SideEngineSample/CustomCountDownViewController.swift
//
//  TestSideEngineViewController.swift
//  SideEngineSample
//
//  Created by Phoenix on 20/04/2022.
//

import UIKit
import AVFoundation
import AudioToolbox
import BBSideEngine

protocol CustomTimerDelegate {
    func didFinishTimer()
}
class CustomCountDownViewController: UIViewController {
  
    var delegate: CustomTimerDelegate?
    
    @IBOutlet var mainTitleLabel : UILabel!
    @IBOutlet var secondsLabel : UILabel!
    @IBOutlet var countDownLabel : UILabel!
    
    //Handle countdown timer
    var counter = BBSideEngineManager.shared.countDownDuration
    var counterTimer : Timer!
 
    var countryCode: String?
    var emergencyContact: String?
    var emergencyContactName: String?
    var emergencyContactEmail: String?
    
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view.
        self.configureTimer()
     }
    
    @IBAction func closeTapped() {
        self.stopTimer(finished: false)
        BBSideEngineManager.shared.resumeSideEngine() //You need to resume side engine when go to back screen
        if isModal == true {
            self.dismiss(animated: true)
        }else {
            self.navigationController?.popViewController(animated: true)
        }
        
    }
    
    func cancelAutoIncident()  {
        if self.counterTimer != nil{
            self.counterTimer.invalidate()
            self.counterTimer = nil
         }
        
        if isModal == true {
            self.dismiss(animated: true)
        }else {
            self.navigationController?.popViewController(animated: true)
        }
    }
    
 
    override func viewWillAppear(_ animated: Bool) {
        navigationController?.setNavigationBarHidden(true, animated: false)
        
    }
 
}


//--------------------------------------------------------//
//MARK: **************UI**************
//--------------------------------------------------------//
extension CustomCountDownViewController {
    //TODO: Configure Timer
    func configureTimer () {
         countDownLabel.text = "\(counter)"
         self.counterTimer = Timer.scheduledTimer(timeInterval: 1, target: self, selector: #selector(self.runTimer), userInfo: nil, repeats: true)
      }
    
    
    //TODO: Support screen
    func openSupportScreen() {
        
        let state = UIApplication.shared.applicationState
        if state != .active || isModal == true {
            
            //If App in the background mode then auto verify incident
            BBSideEngineManager.shared.notifyPartner()
            //Delegate used to send sms or email in user emergency contacts
            if self.delegate != nil {
                self.delegate?.didFinishTimer()
            }
            self.closeTapped() //Auto dismiss and resume side engine
        } else {
            DispatchQueue.main.async {
                let storyBoard : UIStoryboard = UIStoryboard(name: "Main", bundle:nil)
                let controller = storyBoard.instantiateViewController(withIdentifier: "TestSideEngineSupportViewController") as! TestSideEngineSupportViewController
                controller.countryCode = self.countryCode
                controller.emergencyContact = self.emergencyContact
                controller.emergencyContactName = self.emergencyContactName
                controller.emergencyContactEmail = self.emergencyContactEmail
                self.navigationController?.pushViewController(controller, animated: true)
            }
        }
        
    }
 
}

//--------------------------------------------------------//
//MARK: **************Start count down timer**************
//--------------------------------------------------------//
extension CustomCountDownViewController{
    @objc func runTimer() {
        counter = counter - 1
        countDownLabel.text = "\(counter)"
        if counter >= 0 {
            AudioServicesPlayAlertSound(SystemSoundID(kSystemSoundID_Vibrate))
        }
        
        if (counter <= 0) {
            self.stopTimer(finished: true)
        }
    }
    
    func stopTimer(finished : Bool) {
        countDownLabel.text = "\(0)"
        self.counter = 0
        if self.counterTimer != nil{
            self.counterTimer.invalidate()
            self.counterTimer = nil
         }
        
        if finished == true {
            DispatchQueue.main.async {
                self.openSupportScreen()
            }
        }
     }
}


extension UIViewController {

    var isModal: Bool {

        let presentingIsModal = presentingViewController != nil
        let presentingIsNavigation = navigationController?.presentingViewController?.presentedViewController == navigationController
        let presentingIsTabBar = tabBarController?.presentingViewController is UITabBarController

        return presentingIsModal || presentingIsNavigation || presentingIsTabBar
    }
}
